modules/core/src/main/java/org/apache/ignite/internal/future/OrderingFuture.java

```diff
@@ -529,12 +537,14 @@ public void notifyHeadToTail(T result, Throwable exception, ListNode<T> lastNoti
                 stack.addFirst(node);
             }
 
+            NotificationContext context = new NotificationContext();
```

Comment:
Why not just store the exception in a local variable and just pass it to `org.apache.ignite.internal.future.OrderingFuture.DependentAction#onCompletion`?

Reply:
We need to create the exception lazily because not all dependents need it, so it might happen that no dependent needs the exception. Context allows us to implement that lazy creation logic. Please note that it is much more expensive to create an exception that a context object.

Reply:
In a local variable, the exception will also be calculated lazily, I think a class is not needed for this.